File: Data Structures & Algorithms/remove-element/submission-2.go
// func removeElement(nums []int, val int) int {
// 	k := 0
// 	for _, currentValue := range nums {
// 		if currentValue != val {
// 			nums[k] = currentValue
// 			k++
// 		}
// 	}
// 	return k
// }

func removeElement(nums []int, val int) int {
	i, l := 0, len(nums)
	for i < l {
		if nums[i] == val {
			l--
			nums[i] = nums[l]
		} else {
			i++
		}
	}
	return i
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1] * len(nums)
        # 1 2 4 6
        # 1 1 2 8 
        prefix = 1
        for i in range(len(nums)):
            res[i] *= prefix
            prefix *= nums[i]

        #.48   24  12 8
        suffix = 1
        for i in range(len(nums) - 1, -1, -1):
            res[i] *= suffix
            suffix *= nums[i]

        return res

    